### PR TITLE
Only use MasterSecret for local message encryption.

### PR DESCRIPTION
**Pull request by moxie0**

### CODE REVIEW

File: src/org/thoughtcrime/securesms/jobs/MasterSecretDecryptJob.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  public MasterSecretDecryptJob(Context context) {
+    super(context, JobParameters.newBuilder()
+                                .withRequirement(new MasterSecretRequirement(context))
+                                .create());
+  }
+
+  @Override
+  public void onRun(MasterSecret masterSecret) {
+    EncryptingSmsDatabase smsDatabase = DatabaseFactory.getEncryptingSmsDatabase(context);
+    SmsDatabase.Reader    smsReader  = smsDatabase.getDecryptInProgressMessages(masterSecret);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/jobs/MasterSecretDecryptJob.java]: align

File: src/org/thoughtcrime/securesms/database/MmsDatabase.java [hunk 2]
```diff
@@ -563,15 +576,17 @@ public long copyMessageInbox(MasterSecret masterSecret, long messageId) throws M
...
+      return insertMediaMessage(new MasterSecretUnion(masterSecret), request.getPduHeaders(),
                                 request.getBody(), contentValues);
     } catch (NoSuchMessageException e) {
       throw new MmsException(e);
     }
   }
 
-  private Pair<Long, Long> insertMessageInbox(MasterSecret masterSecret, IncomingMediaMessage retrieved,
-                                              String contentLocation, long threadId, long mailbox)
+  private Pair<Long, Long> insertMessageInbox(MasterSecretUnion masterSecret,
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/database/MmsDatabase.java]: Missing some `@NonNull`s here :)

File: src/org/thoughtcrime/securesms/jobs/PushDecryptJob.java [hunk 3]
```diff
@@ -315,55 +342,70 @@ private void handleSynchronizeSentTextMessage(MasterSecret masterSecret,
...
+                                   @NonNull Optional<Long> smsMessageId)
+  {
     EncryptingSmsDatabase smsDatabase = DatabaseFactory.getEncryptingSmsDatabase(context);
 
     if (!smsMessageId.isPresent()) {
-      Pair<Long, Long> messageAndThreadId = insertPlaceholder(masterSecret, envelope);
+      Pair<Long, Long> messageAndThreadId = insertPlaceholder(envelope);
       smsDatabase.markAsLegacyVersion(messageAndThreadId.first);
-      MessageNotifier.updateNotification(context, masterSecret, messageAndThreadId.second);
+      MessageNotifier.updateNotification(context, masterSecret.getMasterSecret().orNull(), messageAndThreadId.second);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/jobs/PushDecryptJob.java]: MessageNotifier shouldn't take in an Optional?

Comment (moxie0) [inline src/org/thoughtcrime/securesms/jobs/PushDecryptJob.java]: Maybe, but it has always taken nullable in the past.  I'm just preserving the behavior.

File: src/org/thoughtcrime/securesms/crypto/storage/TextSecurePreKeyStore.java [hunk 4]
```diff
@@ -129,28 +136,65 @@ public void removeSignedPreKey(int signedPreKeyId) {
...
-    if (recordVersion != CURRENT_VERSION_MARKER) {
+    if (recordVersion > CURRENT_VERSION_MARKER) {
       throw new AssertionError("Invalid version: " + recordVersion);
     }
 
-    return masterCipher.decryptBytes(readBlob(fin));
+    byte[] serializedRecord = readBlob(fin);
+
+    if (recordVersion < PLAINTEXT_VERSION && masterSecret != null) {
+      MasterCipher masterCipher  = new MasterCipher(masterSecret);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/crypto/storage/TextSecurePreKeyStore.java]: extra space sneaking in

File: src/org/thoughtcrime/securesms/database/MmsDatabase.java [hunk 5]
```diff
@@ -474,6 +466,36 @@ public void setAllMessagesRead() {
...
+      type = Types.ENCRYPTION_ASYMMETRIC_BIT;
+    }
+
+    updateMessageBodyAndType(messageId, body, Types.ENCRYPTION_MASK, type);
+  }
+
+  private Pair<Long, Long> updateMessageBodyAndType(long messageId, String body, long maskOff, long maskOn) {
+    SQLiteDatabase db = databaseHelper.getWritableDatabase();
+    db.execSQL("UPDATE " + TABLE_NAME + " SET " + BODY + " = ?, " +
+               MESSAGE_BOX + " = (" + MESSAGE_BOX+ " & " + (Types.TOTAL_MASK - maskOff) + " | " + maskOn + ") " +
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/database/MmsDatabase.java]: found where the extra space came from (after `MESSAGE_BOX`)

**Comment by mcginty** [comment]
placeholder for suggestion of making a `getSymmetric()`/`isSymmetricPresent()`, `getAsymmetric()`/`isAsymmetricPresent()` helper set of methods to reduce the `masterSecret.getMasterSecret().get()` calls to `masterSecret.getSymmetric()`.

**Pull request closed without merging**